habitat/core/dataset.py

```diff
@@ -236,20 +246,116 @@ def get_splits(
+    scene should be loaded consecutively. However, if too many consecutive
+    episodes from same scene are feed into RL model, the model will risk to
+    overfit that scene. Therefore it's better to load same scene consecutively
+    and switch once a number threshold is reached.
+    Currently supports the following features:
+        Cycling: when all episodes are iterated, cycle back to start instead of
+            throwing StopIteration.
+        Cycling with shuffle: when cycling back, shuffle episodes groups
+            grouped by scene.
+        Group by scene: episodes of same scene by grouped and will be loaded
+            consecutively.
+        Set max scene repeat: set a number threshold on how many episodes from
+        the same scene can be loaded consecutively.
+        Sample episodes: sample the specified number of episodes.
+    """
+
+    def __init__(
+        self,
+        episodes: List[T],
+        cycle: bool = True,
+        shuffle_scene_when_cycle: bool = False,
+        group_by_scene: bool = True,
+        max_scene_repeat: int = -1,
+        num_episode_sample: int = -1,
+    ):
+        r"""
+        Initialize episode iterator,
         Args:
-            num_episodes: number of episodes to sample, input -1 to use
-            whole episodes
+            episodes: list of episodes.
+            cycle: if true, cycle back to first episodes when StopIteration.
+            shuffle_scene_when_cycle: if true, shuffle scene groups when cycle.
+            group_by_scene: if true, group episodes from same scene.
+            max_scene_repeat: threshold of how many episodes from the same
+                scene can be loaded consecutively. -1 for no limit
+            num_episode_sample: number of episodes to be sampled.
+                -1 for no sampling.
         """
-        if num_episodes == -1:
-            return
-        if num_episodes < -1:
-            raise ValueError(
-                f"Invalid number for episodes to sample: {num_episodes}"
+        # sample episodes
+        if num_episode_sample != -1:
+            if num_episode_sample < -1:
+                raise ValueError(
+                    f"Invalid episode number to sample: {num_episode_sample}"
+                )
+            episodes = np.random.choice(
+                episodes, num_episode_sample, replace=False
             )
-        self.episodes = np.random.choice(
-            self.episodes, num_episodes, replace=False
-        )
+        self.episodes = episodes
+        self.cycle = cycle
+        self.group_by_scene = group_by_scene
+        if group_by_scene:
+            num_scene_groups = len(
+                list(groupby(episodes, key=lambda x: x.scene_id))
+            )
+            num_unique_scenes = len(set([e.scene_id for e in episodes]))
+            if num_scene_groups >= num_unique_scenes:
+                self.episodes = sorted(self.episodes, key=lambda x: x.scene_id)
```

Comment:
Probably makes sense to always sort.  The only times when you'd save any time is when `len(self.episodes)` is short enough that sorting doesn't take meaningful time.  In the general case where you need to sort, the checks to see if you need to sort will take non-trivial time.

Also, the sorting algo python uses is `O(n)` when the list is already sorted, so there isn't even any savings there!

Reply:
@erikwijmans one scenario that might be worth considering: episodes are grouped by scene already but the scenes themselves are not in order (i.e. episodes shuffled by scene): in this case sorting is unnecessary and will take `O(nlgn)`.

Reply:
Better plan.  Use `pydash.group_by` as that doesn't have the limitation that things need to be sorted! https://pydash.readthedocs.io/en/latest/api.html#pydash.collections.group_by

Almost certainly some dependency already relies on pydash

Reply:
pydash also provides some cool syntactic sugar, so `num_scene_groups` would be:
```
num_scene_groups = len(pydash.group_by(episodes, 'scene_id'))
```

Reply:
I checked with `pipdeptree`, `pydash` is not currently replied on. I'll be conservative and not using `pydash` for now, we can definitely discuss this more though.